src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1426,10 +1426,81 @@ public static unsafe nuint NarrowUtf16ToAscii(char* pUtf16Buffer, byte* pAsciiBu
             goto Finish;
         }
 
-        private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pAsciiBuffer, nuint elementCount)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool VectorContainsNonAsciiChar(Vector128<ushort> utf16Vector, Vector128<short> asciiMaskForTestZ, Vector128<ushort> asciiMaskForAddSaturate)
         {
-            // This method contains logic optimized for both SSE2 and SSE41. Much of the logic in this method
-            // will be elided by JIT once we determine which specific ISAs we support.
+            if (Sse2.IsSupported)
+            {
+                // See comments in GetIndexOfFirstNonAsciiChar_Sse2 for information about how this works.
+
+                if (Sse41.IsSupported)
+                {
+                    if (!Sse41.TestZ(utf16Vector.AsInt16(), asciiMaskForTestZ))
+                    {
+                        return true;
+                    }
+                }
+                else
+                {
+                    if ((Sse2.MoveMask(Sse2.AddSaturate(utf16Vector, asciiMaskForAddSaturate).AsByte()) & 0b_1010_1010_1010_1010) != 0)
+                    {
+                        return true;
+                    }
+                }
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                Vector128<ushort> maxChars = AdvSimd.Arm64.MaxPairwise(utf16Vector, utf16Vector);
+                if ((maxChars.AsUInt64().ToScalar() & 0xFF80FF80FF80FF80) != 0)
```

Comment:
Just thinking aloud... For the xplat logic we have two options for handling such scenarios:
1. Expose `MaxPairwise` like APIs and have the JIT import as `Shuffle + Max` on x86/x64
2. Recognize the `Shuffle + Max` pattern on other platforms like ARM in morph and normalize to `MaxPairwise`

Both of these seem like viable options with 1 being a bit more user friendly. I'll probably open an issue covering this for .NET 8

Reply:
Thanks, sounds good. Marking this conversation resolved.

Reply:
I remember that `.ToScalar()` would move data from vector to GPR. Can we somehow do it without converting to GPR and using `Vector.Zero` or similar?

Reply:
Doing `(maxChars & Vector128.Create(0xFF80FF80FF80FF80, ulong.MaxValue)) != 0` should work. We can also (but don't currently) optimize that to just `TestZ` (I think we'll emit `And, TestZ` on x64 today).

Reply:
It seems, we cannot avoid SIMD to GPR transfer easily.

[1] Current version
`if ((maxChars.AsUInt64().ToScalar() & 0xFF80FF80FF80FF80) != 0)`
```
umaxp   v18.8h, v16.8h, v16.8h      # get maxChars
umov    x0, v18.d[0]
tst     x0, #0xd1ffab1e
```
[Full asm dump](https://gist.github.com/SwapnilGaikwad/a1d5e6481cd2fea948647303bcad4ee5)

[2] Above
`if ((maxChars.AsUInt64() & Vector128.Create(0xFF80FF80FF80FF80, 0)) != 0) -> compiler error`
with `(maxChars.AsUInt64() & Vector128.Create(0xFF80FF80FF80FF80, 0)) != 0` complains that vector cannot be compared to int.
`Operator '!=' cannot be applied to operands of type 'Vector128<ulong>' and 'int'`

[3] Compare to vector zero
`if ((maxChars.AsUInt64() & Vector128.Create(0xFF80FF80FF80FF80)) != Vector128<ulong>.Zero)`

```
umaxp   v19.8h, v16.8h, v16.8h
and     v18.2d, v19.2d, v18.2d
umaxv   s18, v18.4s
umov    w0, v18.s[0]
cmp     w0, #0
```
[Full asm dump](https://gist.github.com/SwapnilGaikwad/dbeda0e4614bfe265ac02f83025b612a)

As one would expect, comparison to vector slows the execution.

```
|   Method |        Job                                                                                                 Toolchain | size | encName |        Mean |    Error |   StdDev |      Median |         Min |         Max | Ratio | MannWhitney(2%) |  Gen 0 | Allocated | Alloc Ratio |
|--------- |----------- |---------- --------------------------------------------------------------------------------------------- |----- |-------- |------------:|---------:|---------:|------------:|------------:|------------:|------:|---------------- |-------:|----------:|------------:|
| GetBytes | Job-LHAUYZ |          /curr_patch/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |    53.66 ns | 0.040 ns | 0.038 ns |    53.66 ns |    53.60 ns |    53.73 ns |  1.00 |            Base | 0.0764 |      40 B |        1.00 |
| GetBytes | Job-SHUITZ | /compare_to_vec_zero/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii |    55.30 ns | 0.421 ns | 0.394 ns |    55.06 ns |    54.91 ns |    56.21 ns |  1.03 |          Slower | 0.0764 |      40 B |        1.00 |
|          |            |                                                                                                         |      |         |             |          |          |             |             |             |       |                 |        |           |             |
| GetBytes | Job-LHAUYZ |          /curr_patch/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii |   302.51 ns | 0.340 ns | 0.302 ns |   302.42 ns |   302.15 ns |   303.19 ns |  1.00 |            Base | 1.0236 |     536 B |        1.00 |
| GetBytes | Job-SHUITZ | /compare_to_vec_zero/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii |   326.41 ns | 3.055 ns | 2.858 ns |   326.88 ns |   320.60 ns |   330.57 ns |  1.08 |          Slower | 1.0245 |     536 B |        1.00 |
|          |            |                                                                                                         |      |         |             |          |          |             |             |             |       |                 |        |           |             |
| GetBytes | Job-LHAUYZ |          /curr_patch/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun | 2048 |   ascii | 1,096.45 ns | 2.614 ns | 2.445 ns | 1,097.77 ns | 1,093.27 ns | 1,099.60 ns |  1.00 |            Base | 3.9500 |    2072 B |        1.00 |
| GetBytes | Job-SHUITZ | /compare_to_vec_zero/bin/testhost/net7.0-Linux-Release-arm64/shared/Microsoft.NETCore.App/7.0.0/corerun | 2048 |   ascii | 1,184.38 ns | 4.176 ns | 3.702 ns | 1,185.09 ns | 1,174.93 ns | 1,187.76 ns |  1.08 |          Slower | 3.9504 |    2072 B |        1.00 |
```

Reply:
I think its fine to keep it as `if ((maxChars.AsUInt64().ToScalar() & 0xFF80FF80FF80FF80) != 0)` for now. We should ensure an issue is logged to improve the codegen around `if ((maxChars.AsUInt64() & Vector128.Create(0xFF80FF80FF80FF80, 0)) != Vector128<ulong>.Zero)` however.

Reply:
Agreed. I'll keep this as is then.